### PR TITLE
PACA-916: Use Project Reactor for concurrent sink

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-916](https://jira.cms.gov/browse/PACA-916)

**User Story or Bug Summary:**

This PR replaces the message queue and executor service implementation of `ConcurrentRdaSink` with a simpler one based on [Project Reactor](https://projectreactor.io) fluxes.  All existing functionality of the concurrent sink remains the same.  The reactive implementation is simpler and easier to maintain.

**NOTE:** This PR introduces a new maven dependency into the pipeline application:

```
<dependency>
        <!-- Functional-reactive programming api for concurrent claim ingestion. -->
        <groupId>io.projectreactor</groupId>
        <artifactId>reactor-core</artifactId>
</dependency>
```

---

### What Does This PR Do?

Overview of changes in this PR:

- Uses project reactor to simplify ConcurrentRdaSink and associated classes.
- Adds additional logging to RDA pipeline jobs to specify reasons for shutting down.
- Adds additional logging to allow cause of shutdown delays (if any) when jobs are interrupted.
- Fixes bug that caused processed count to be lost if pipeline exited due to an exception.

Some of the changes are hard to notice in the github diffs:

- `SequenceNumberWriter` is a renaming and simplification of `SequenceNumberWriterThread`.  The name changed because the object is no longer intended to be submitted as a thread in an executor service.
- Ditto for `ClaimWriter` and `ClaimWriterThread`.
- `WriterThreadPool` was no longer needed and some of its logic was moved into `ConcurrentRdaSink` rather than being delegated as before.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [x] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by KyRobbins — APPROVED** [review]

**Review by StewGoin — APPROVED** [review]
Ok by me.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java [hunk 1]
```diff
@@ -153,16 +277,39 @@ public int writeClaims(Collection<TClaim> objects) throws ProcessingException {
...
+      closer.close(sink::close);
+      log.info("shutdown check for errors");
+      closer.close(this::throwIfErrorPresent);
+      log.info("shutdown finish");
+      closer.finish();
+    } catch (ProcessingException ex) {
+      log.error("shutdown failed: ex={}", ex.getMessage(), ex);
+      throw ex;
     } catch (Exception ex) {
+      log.error("shutdown failed: ex={}", ex.getMessage(), ex);
```
Comment (daryll) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java]: lots of steps for the shutdown.
i assume it must always happen in this sequence.
do we need to add any documentation to this method if these are done out of sequence

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java [hunk 2]
```diff
@@ -172,6 +319,83 @@ public void shutdown(Duration waitTime) throws ProcessingException {
...
+   *
+   * @param waitTime how long to wait for the latch
+   * @throws Exception if any exception caused the wait to fail
+   */
+  private void waitForLatch(Duration waitTime) throws Exception {
+    final long startMillis = System.currentTimeMillis();
+    InterruptedException error = null;
+    boolean successful = false;
+    for (int i = 1; i <= 10; ++i) {
+      try {
```
Comment (daryll) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java:387]: wondering if we should we make this a constant explaining what this is ?
its hard to know why we would use 10 vs. any other number here

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/StandardGrpcRdaSource.java [hunk 3]
```diff
@@ -220,20 +220,24 @@ public int retrieveAndProcessObjects(int maxPerBatch, RdaSink<TMessage, TClaim>
...
             if (isUnexpectedDroppedConnectionException(lastProcessedTime, ex)) {
               processResult.setException(ex);
             }
           } catch (ProcessingException ex) {
+            log.info("shutting down due to ProcessingException: {}", ex.getMessage());
             flushBatch = false;
             processResult.addCount(ex.getProcessedCount());
             processResult.setException(ex);
           } catch (Exception ex) {
+            log.info("shutting down due to Exception: {}", ex.getMessage());
```
Comment (daryll) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/StandardGrpcRdaSource.java:239]: many thx for the logging here !

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java [hunk 4]
```diff
@@ -33,17 +121,55 @@
    */
   public ConcurrentRdaSink(
       int maxThreads, int batchSize, Supplier<RdaSink<TMessage, TClaim>> sinkFactory) {
```
Comment (daryll) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java:121]: great work on making this so compact !

would be cool if we can add anything here to help future devs decipher the schedulers, publisher, processors...etc.

**Review by daryll — APPROVED** [review]
lgtm!

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java]: Closer does everything synchronously and the first thing that happens is waiting for the latch to signal the fluxes have shut down.  At that point its just cleanup.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java]: I have refactored the method to remove the need for this counter.

**Review by KyRobbins — DISMISSED** [review]

**Review by daryll — DISMISSED** [review]
looks good...thk you

**Review by lsmitchell — COMMENTED** [review]
Overall looks more straightforward and less code than the previous implementation. Thanks for the detailed documentation (and improving my initial documentation attempts in some places); made following this much easier!

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java [hunk 5]
```diff
@@ -69,10 +191,83 @@ public static <TMessage, TClaim> RdaSink<TMessage, TClaim> createSink(
...
+    final var claimPartitioner = new StringPartitioner<>(claimWriters);
+
+    // Flux used to trigger idle checks.
+    // The interval flux will emit a time in milliseconds every time it fires.
+    final var idleTimerFlux =
+        Flux.interval(IdleCheckInterval, claimWriterScheduler)
+            // Stop sending messages when shutdown sets the running flag to false.
+            .takeWhile(o -> isRunning())
+            // Drop any extra messages if the writer is currently busy.
+            .onBackpressureLatest()
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java:215]: Good question.  Think of these idle messages as reminders for the writer to flush the database if it's idle. If the writer is busy when one of those reminders is due then clearly it's not idle so the reminder is redundant.  There will be another one coming along a little while later.  It helps with throughput to just drop the redundant one.  I expect the writers to be busy more often than they are not

[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java:215]: No claims can be lost if these idle messages are dropped.  Partial buffers are also written when the stream closes.  If claims are arriving fast enough these idle messages are not necessary.

[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java:215]: Gotcha, so these messages are just reminder messages, not the claims themselves. I think that's what I wasn't getting here. Thanks for explaining!

**Review by lsmitchell — APPROVED** [review]

**Review by KyRobbins — APPROVED** [review]

**Pull request merged**